lib/grp.gi

```diff
@@ -130,14 +130,95 @@ end);
     TryNextMethod();
   else
-    Error(
-  "`MinimalGeneratingSet' currently assumes that the group is solvable, or\n",
-  "already possesses a generating set of size 2.\n",
-  "In general, try `SmallGeneratingSet' instead, which returns a generating\n",
-  "set that is small but not of guaranteed smallest cardinality");
+    cs := ChiefSeries(G);
+    phi_GbyG1 := NaturalHomomorphismByNormalSubgroup(G,cs[2]);
+    GbyG1 := ImagesSource(phi_GbyG1);
+    mingenset_k_reps := List(SmallGeneratingSet(GbyG1), x -> PreImagesRepresentative(phi_GbyG1, x));
+    for k in [3..Length(cs)] do # Lifting
+      mingenset_km1_reps := mingenset_k_reps;
+      Gk := cs[k];
+      Gkm1 := cs[k-1];
+      phi_GbyGk := NaturalHomomorphismByNormalSubgroup(G,Gk);
+      phi_Gkm1byGk := NaturalHomomorphismByNormalSubgroup(Gkm1,Gk);
+      Gkm1byGk := ImagesSource(phi_Gkm1byGk);
+      GbyGk := ImagesSource(phi_GbyGk);
+      check := gx -> GbyGk = GroupByGenerators(ImagesSet(phi_GbyGk,gx));
+      Gkm1byGk_elem_reps := List(AsList(Gkm1byGk),x -> PreImagesRepresentative(phi_Gkm1byGk,x));
+      Gkm1byGk_gen := SmallGeneratingSet(Gkm1byGk);
+      Gkm1byGk_gen_reps := List(Gkm1byGk_gen,x -> PreImagesRepresentative(phi_Gkm1byGk,x));
+      g := ShallowCopy(mingenset_km1_reps);
+      stop := false;
+      if IsAbelian(Gkm1byGk) then
+        if check(g) then mingenset_k_reps := g; fi;
+        for i in [1..Length(g)] do 
+          if stop then break; fi;
+          for j in [1..Length(Gkm1byGk_gen_reps)] do
+            temp := g[i];
+            g[i] := temp * Gkm1byGk_gen_reps[j];
+            if check(g) then 
+              mingenset_k_reps := g;
+              stop := true;
+              break;
+            fi;
+            g[i] := temp;
+          od;
+        od;
+        Add(g,Gkm1byGk_gen_reps[1]);
+        mingenset_k_reps := g;
+      else
+          g0 := ShallowCopy(mingenset_km1_reps);
+          g := ShallowCopy(mingenset_km1_reps);
+          Add(g,Gkm1byGk_elem_reps[1]);
+          g1 := ShallowCopy(g);
+          for g in [g0,g1] do 
+            if stop then break;fi;
+            l := Length(g);
+            L := Length(Gkm1byGk_elem_reps);
+            s := L^l;
+            last := [];
+            for i in [l,l-1..1] do last[i] := 1; od;
+            gmod := ShallowCopy(g);
+            for x in [0..s-1] do 
+              xl := [];
+              for i in [1..l] do xl[i] := 0; od;
+              i := 1;
+              while x > 0 do 
+                r := RemInt(x,L);
+                x := QuoInt(x,L);
+                xl[i]:=r;
+                i:= i+1;
+              od;
+              for i in [1..l] do 
+                if xl[i] <> last[i] then
+                  gmod[i] := g[i] * Gkm1byGk_elem_reps[xl[i]+1];
+                fi;
+              od;
+              if check(gmod) then
+                mingenset_k_reps := gmod;
+                stop := true;
+                break;
+              fi;
+              last := xl;
```

Comment:
`last` is a global variable in GAP (like `_` in the python interpreter), so maybe this isn't a good name for this variable?

Reply:
I changed it now.